### PR TITLE
[AST][Sema] Fixes the IterativeTypeChecker and better manages circular protocol inheritance

### PR DESCRIPTION
**Pull request by manavgabhawala**
#### What's in this pull request?

Improves the diagnostics produced for circular protocol inheritance, further it improves the IterativeTypeChecker to use loops instead of recursion to help minimize the stack size and lastly improves the efficiency of detecting circular inheritance by making it a part of the main cycle when checking for inherited protocols.

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

 **Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| OS X platform | @swift-ci Please test OS X platform |
| Linux platform | @swift-ci Please test Linux platform |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by tkremenek** [comment]
@swift-ci test

**Comment by tkremenek** [comment]
@jopamer Please review.

**Comment by tkremenek** [comment]
@swift-ci test

**Comment by jopamer** [comment]
LGTM.

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
Awesome work @manavgabhawala, it's really great to see the iterative decl checker getting some attention.

**Comment by manavgabhawala** [comment]
The linux build timed out, and I'm not really sure why `SourceKit/DocSupport/doc_clang_module.swift ` and `SourceKit/DocSupport/doc_swift_module.swift `. Any help to figure out how I can fix them would be great.

**Comment by DougGregor** [comment]
@swift-ci Please test and merge

**Comment by manavgabhawala** [comment]
@DougGregor I think the builds timed out. Is there something I need to do to fix it?

**Comment by CodaFi** [comment]
@harlanhaskins @DougGregor @jopamer Ping.

This patch needs to be retested.

**Comment by harlanhaskins** [comment]
@swift-CI please test

**Comment by harlanhaskins** [comment]
@shahmishal This build keeps timing out, and I have no idea why...

**Comment by CodaFi** [comment]
Perhaps a rebase would do it (it is quite old).  @manavgabhawala Can you `git rebase upstream master`?

**Comment by manavgabhawala** [comment]
@CodaFi done

**Comment by harlanhaskins** [comment]
@swift-ci please test this pull request

**Comment by harlanhaskins** [comment]
@swift-ci please test

**Comment by manavgabhawala** [comment]
@harlanhaskins any update on why its timing out?

**Comment by shahmishal** [comment]
@swift-ci please test

File: lib/AST/ConformanceLookupTable.cpp [hunk 1]
```diff
@@ -458,7 +458,9 @@ bool ConformanceLookupTable::addProtocol(NominalTypeDecl *nominal,
 
       case ConformanceEntryKind::Implied:
         // An implied conformance is better than a synthesized one.
-        if (kind == ConformanceEntryKind::Synthesized)
+        // Ignore implied ciruclar protocol inheritance
```
Comment (slavapestov) [inline lib/AST/ConformanceLookupTable.cpp]: Typo -- "ciruclar"

**Comment by slavapestov** [comment]
@swift-ci Please test and merge

**Comment by rintaro** [comment]
Tested locally,
Just reverting changes in `IterativeTypeChecker.cpp` solves the problem.
```
$ git fetch origin pull/2123/head:PR-2123
$ git checkout PR-2123
$ git rebase master

# Resolve a conflict (caused by `defer` => `SWIFT_DEFER` naming change)
# in IterativeTypeChecker.cpp

$ utils/build-script -RT

# Confirmed it never finish
# validation-test/compiler_crashers/28261-swift-iterativetypechecker-satisfy.swift
# validation-test/SIL/crashers/032-swift-iterativetypechecker-satisfy.sil
^C

$ git checkout master -- lib/Sema/IterativeTypeChecker.cpp
$ utils/build-script -RT

<success>

$
```

**Comment by CodaFi** [comment]
@rintaro I can shepherd these changes in with another pull request.

**Comment by CodaFi** [comment]
Superseded by the merge of #3747.  Thanks @manavgabhawala!

**Pull request closed without merging**